analysis.py

```diff
@@ -0,0 +1,207 @@
+#!/usr/bin/env python3
+
+import argparse
+import os
+import random
+import time
+import datetime
+from typing import Tuple
+
+# this script assumes you installed deltachat: py.delta.chat/install.html
+import deltachat
+from plugins import EchoPlugin, ReceivePlugin, parse_msg
+from output import Output
+
+
+def setup_account(output: object, entry: dict, data_dir: str, plugin: object) -> deltachat.Account:
+    """Creates a Delta Chat account for a given credentials dictionary.
+
+    :param output: the Output object which takes care of the results
+    :param data_dir: the directory where the accounts the argparse arguments
+        passed to this script
+    :param entry: a dictionary with at least an "addr" and a "app_pw" key
+    :param plugin: a plugin class which the bot will use
+    """
+    assert entry.get("addr") and entry.get("app_pw")
+
+    begin = time.time()
+    # create deltachat.Account
+    os.mkdir(os.path.join(data_dir, entry["addr"]))
+    db_path = os.path.join(data_dir, entry["addr"], "db.sqlite")
+
+    ac = deltachat.Account(db_path)
+    #ac.add_account_plugin(deltachat.events.FFIEventLogger(ac))
+    ac.add_account_plugin(plugin)
+    ac.set_config("addr", entry["addr"])
+    ac.set_config("mail_pw", entry["app_pw"])
+
+    for name in ("send_server", "mail_server"):
+        val = entry.get(name)
+        if val is not None:
+            ac.set_config(name, val)
+
+    ac.set_config("mvbox_move", "0")
+    ac.set_config("mvbox_watch", "0")
+    ac.set_config("sentbox_watch", "0")
+    ac.set_config("bot", "1")
+    configtracker = ac.configure()
+    configtracker.wait_finish()
+    ac.start_io()
+    duration = time.time() - begin
+    print("%s: successful login as %s in %.1f seconds." % (entry["addr"], plugin.name, duration))
+    if plugin is not EchoPlugin:
+        output.submit_login_result(entry.get("addr"), duration)
+    ac.output = output
```

Comment:
I did this because I thought you could only add classes as plugins, not objects; and the output object only makes sense if there is one object to store all the information to be written to the output .csv file.

Now with the FFIEventLogger I know that you could also add objects, and that it's just the same over all of my different bot instances.

But right now this does the job, and in some ways it's easier than if the output was an account plugin; I don't know how easy it is to call a specific plugin from an account object. It's only used in https://github.com/deltachat/performance-analysis/blob/559446a402199689540f9330bbbbc1bf248ff50b/plugins.py#L32 at the moment.

Reply:
it's ok for now. indeed using static methods is a bit weird.  Usually a plugin gets instantiated and has attribute where it can store/track state.

Reply:
Yes, the static method was introduced by https://github.com/deltachat/bg-deltachat/pull/162, as working solution for a problem I didn't fully understand. Are there alternatives?

Reply:
Fixed in https://github.com/deltachat/eppdperf/pull/7/commits/e25769abe2ec60f3c23eb65faead1bd7767a9e03 :)